File: blast/blast-stress-solver/src/tests/solver-mechanisms.test.ts
/**
 * Empirical tests of two physics mechanisms at the JS SOLVER level — the mirror of the Rust
 * `tests/solver_mechanisms_test.rs`. We assert the actual numeric behavior rather than
 * trusting that the feature is wired up, so we can be confident BOTH libraries' solvers
 * implement orientation-dependent gravity and excess-force-on-fracture.
 *
 * Whether each library's high-level destruction pipeline FEEDS these mechanisms correctly is
 * a separate, integration-level question (see gaps #7/#8 in blast/TESTING.md). Notably, the
 * JS pipeline DOES rotate gravity per actor (destructible-core.ts), while the Rust pipeline
 * does not; conversely the Rust pipeline applies excess forces while the JS pipeline does not.
 */
import { describe, it, expect } from 'vitest';
import { existsSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import type * as Runtime from '..';

const here = dirname(fileURLToPath(import.meta.url));
const runtimeAvailable = existsSync(resolve(here, '../../dist/stress_solver.wasm'));

type Vec3 = { x: number; y: number; z: number };
type Node = { centroid: Vec3; mass: number; volume: number };
type Bond = { centroid: Vec3; normal: Vec3; area: number; node0: number; node1: number };

async function loadRuntime() {
  return (await import('../../dist/index.js')) as typeof Runtime;
}

/** 6-node horizontal cantilever along +X, node 0 fixed (mass 0). */
function cantilever(): { nodes: Node[]; bonds: Bond[] } {
  const n = 6;
  const nodes: Node[] = [];
  const bonds: Bond[] = [];
  for (let i = 0; i < n; i += 1) {
    nodes.push({ centroid: { x: i, y: 0, z: 0 }, mass: i === 0 ? 0 : 1, volume: 1 });
  }
  for (let i = 0; i < n - 1; i += 1) {
    bonds.push({
      centroid: { x: i + 0.5, y: 0, z: 0 },
      normal: { x: 1, y: 0, z: 0 },
      area: 0.5,
      node0: i,
      node1: i + 1,
    });
  }
  return { nodes, bonds };
}

// Weak tension/shear, very strong compression: a bending (perpendicular) load snaps the
// beam; an axial load only compresses it.
const bendSensitive = {
  maxSolverIterationsPerFrame: 25,
  compressionElasticLimit: 1e5,
  compressionFatalLimit: 1e6,
  tensionElasticLimit: 0.05,
  tensionFatalLimit: 0.1,
  shearElasticLimit: 0.05,
  shearFatalLimit: 0.1,
};

describe.skipIf(!runtimeAvailable)('Solver mechanisms (requires WASM build)', () => {
  async function overstressForGravity(gravity: Vec3, perActor: boolean): Promise<number> {
    const rt = await (await loadRuntime()).loadStressSolver();
    const { nodes, bonds } = cantilever();
    const solver = rt.createExtSolver({ nodes, bonds, settings: bendSensitive });
    if (perActor) {
      const [actor] = solver.actors();
      expect(actor).toBeDefined();
      expect(solver.addActorGravity(actor!.actorIndex, gravity)).toBe(true);
    } else {
      solver.addGravity(gravity);
    }
    solver.update();
    const count = solver.overstressedBondCount();
    solver.destroy();
    return count;
  }

  it('addGravity is direction sensitive (bending vs axial)', async () => {
    const bending = await overstressForGravity({ x: 0, y: -50, z: 0 }, false);
    const axial = await overstressForGravity({ x: -50, y: 0, z: 0 }, false);
    expect(bending).toBeGreaterThan(axial);
  });

  it('addActorGravity is direction sensitive (the API the JS pipeline uses for rotation)', async () => {
    const bending = await overstressForGravity({ x: 0, y: -50, z: 0 }, true);
    const axial = await overstressForGravity({ x: -50, y: 0, z: 0 }, true);
    expect(bending).toBeGreaterThan(axial);
  });

  it('getExcessForces reports the released load after a bond fractures', async () => {
    const rt = await (await loadRuntime()).loadStressSolver();
    // A 10 kg mass hung from a fixed support by one bond under g = 100 carries ~1000 N.
    const nodes: Node[] = [
      { centroid: { x: 0, y: 1, z: 0 }, mass: 0, volume: 1 },
      { centroid: { x: 0, y: 0, z: 0 }, mass: 10, volume: 1 },
    ];
    const bonds: Bond[] = [
      { centroid: { x: 0, y: 0.5, z: 0 }, normal: { x: 0, y: 1, z: 0 }, area: 1, node0: 0, node1: 1 },
    ];
    const solver = rt.createExtSolver({
      nodes,
      bonds,
      settings: {
        maxSolverIterationsPerFrame: 25,
        compressionElasticLimit: 1, compressionFatalLimit: 2,
        tensionElasticLimit: 1, tensionFatalLimit: 2,
        shearElasticLimit: 1, shearFatalLimit: 2,
      },
    });
    solver.addGravity({ x: 0, y: -100, z: 0 });
    solver.update();
    expect(solver.overstressedBondCount()).toBe(1);

    const perActor = solver.generateFractureCommandsPerActor();
    const events = solver.applyFractureCommands(perActor);
    expect(events.length).toBeGreaterThan(0);

    const freed = solver.actors().find((a) => a.nodes.includes(1));
    expect(freed).toBeDefined();
    const excess = solver.getExcessForces(freed!.actorIndex, { x: 0, y: 0, z: 0 });
    expect(excess).not.toBeNull();
    const mag = Math.hypot(excess!.force.x, excess!.force.y, excess!.force.z);
    // Near-zero would mean the momentum-transfer mechanism is dormant.
    expect(mag).toBeGreaterThan(100);
    solver.destroy();
  });
});

File: blast/blast-stress-solver-rs/tests/solver_mechanisms_test.rs
//! Empirical tests of two physics mechanisms at the SOLVER level (no Rapier, no trust in
//! code inspection): orientation-dependent gravity and excess-force-on-fracture. The JS
//! suite has a mirror of these (`src/tests/solver-mechanisms.test.ts`) so we can be
//! confident BOTH libraries' solvers actually implement them.
//!
//! These verify the *mechanisms exist and respond correctly*. Whether each library's
//! high-level destruction pipeline actually FEEDS them the right inputs is a separate,
//! integration-level question (see gap #7 / gap #8 in blast/TESTING.md and
//! `gravity_orientation_test.rs`).

use blast_stress_solver::*;

/// 6-node horizontal cantilever: node 0 fixed (mass 0), nodes 1..5 along +X.
fn cantilever() -> (Vec<NodeDesc>, Vec<BondDesc>) {
    let n = 6u32;
    let mut nodes = Vec::new();
    let mut bonds = Vec::new();
    for i in 0..n {
        let mass = if i == 0 { 0.0 } else { 1.0 };
        nodes.push(NodeDesc { centroid: Vec3::new(i as f32, 0.0, 0.0), mass, volume: 1.0 });
    }
    for i in 0..n - 1 {
        let (a, b) = (nodes[i as usize].centroid, nodes[(i + 1) as usize].centroid);
        bonds.push(BondDesc { node0: i, node1: i + 1, centroid: (a + b) * 0.5, normal: Vec3::new(1.0, 0.0, 0.0), area: 0.5 });
    }
    (nodes, bonds)
}

/// Weak tension/shear, very strong compression: a bending (perpendicular) load snaps the
/// beam; an axial load only compresses it.
fn bend_sensitive_settings() -> SolverSettings {
    SolverSettings {
        compression_elastic_limit: 1.0e5, compression_fatal_limit: 1.0e6,
        tension_elastic_limit: 0.05, tension_fatal_limit: 0.1,
        shear_elastic_limit: 0.05, shear_fatal_limit: 0.1,
        ..SolverSettings::default()
    }
}

fn overstress_for_gravity(gravity: Vec3, per_actor: bool) -> u32 {
    let (nodes, bonds) = cantilever();
    let mut solver = ExtStressSolver::new(&nodes, &bonds, &bend_sensitive_settings()).unwrap();
    if per_actor {
        let actor = solver.actors()[0].actor_index;
        assert!(solver.add_actor_gravity(actor, gravity), "add_actor_gravity should succeed");
    } else {
        solver.add_gravity(gravity);
    }
    solver.update();
    solver.overstressed_bond_count()
}

/// Mechanism 1a: the GLOBAL gravity vector's direction changes the stress. A perpendicular
/// (bending) load overstresses the beam; an axial load (along the beam, into the anchor)
/// does not.
#[test]
fn add_gravity_is_direction_sensitive() {
    let bending = overstress_for_gravity(Vec3::new(0.0, -50.0, 0.0), false);
    let axial = overstress_for_gravity(Vec3::new(-50.0, 0.0, 0.0), false);
    assert!(
        bending > axial,
        "gravity direction must affect stress: perpendicular bending overstressed {bending} bonds, \
         axial overstressed {axial} (expected bending > axial)"
    );
}

/// Mechanism 1b: the PER-ACTOR gravity API (`add_actor_gravity`) — the exact call the JS
/// pipeline uses with gravity rotated into each actor's local frame — is direction
/// sensitive. This is what makes orientation-correct gravity possible.
#[test]
fn add_actor_gravity_is_direction_sensitive() {
    let bending = overstress_for_gravity(Vec3::new(0.0, -50.0, 0.0), true);
    let axial = overstress_for_gravity(Vec3::new(-50.0, 0.0, 0.0), true);
    assert!(
        bending > axial,
        "add_actor_gravity must be direction sensitive: bending overstressed {bending} bonds, \
         axial overstressed {axial} (expected bending > axial)"
    );
}

/// Mechanism 2: when a loaded bond breaks, `getExcessForces` reports the released load (so a
/// pipeline can throw the freed fragment). A 10 kg mass hung from a fixed support by one
/// bond under g = 100 carries ~1000 N; after the bond fractures the freed mass's actor must
/// report a non-trivial excess force. (The pre-existing solver test only checked finiteness,
/// so this magnitude was never actually asserted.)
#[test]
fn excess_force_reports_released_load() {
    let nodes = vec![
        NodeDesc { centroid: Vec3::new(0.0, 1.0, 0.0), mass: 0.0, volume: 1.0 },
        NodeDesc { centroid: Vec3::new(0.0, 0.0, 0.0), mass: 10.0, volume: 1.0 },
    ];
    let bonds = vec![BondDesc {
        centroid: Vec3::new(0.0, 0.5, 0.0), normal: Vec3::new(0.0, 1.0, 0.0), area: 1.0, node0: 0, node1: 1,
    }];
    let settings = SolverSettings {
        compression_elastic_limit: 1.0, compression_fatal_limit: 2.0,
        tension_elastic_limit: 1.0, tension_fatal_limit: 2.0,
        shear_elastic_limit: 1.0, shear_fatal_limit: 2.0,
        ..SolverSettings::default()
    };
    let mut solver = ExtStressSolver::new(&nodes, &bonds, &settings).unwrap();
    solver.add_gravity(Vec3::new(0.0, -100.0, 0.0));
    solver.update();
    assert_eq!(solver.overstressed_bond_count(), 1, "the loaded bond should overstress");

    let cmds = solver.generate_fracture_commands();
    let events = solver.apply_fracture_commands(&cmds);
    assert!(!events.is_empty(), "the bond should fracture and split the actor");

    // The freed mass's actor must report a meaningful released load (~weight = 1000 N).
    let freed = solver
        .actors()
        .into_iter()
        .find(|a| a.nodes.contains(&1))
        .expect("freed-mass actor exists");
    let (force, _torque) = solver
        .get_excess_forces(freed.actor_index, Vec3::new(0.0, 0.0, 0.0))
        .expect("excess force available right after fracture");
    let mag = force.magnitude();
    assert!(
        mag > 100.0,
        "excess force should report the released load (~1000 N), got |f| = {mag:.1} N \
         ({force:?}) — near-zero would mean the momentum-transfer mechanism is dormant"
    );
}
